core/node/node_sync/src/tree_data_fetcher/provider/mod.rs

```diff
@@ -195,12 +252,36 @@ impl TreeDataProvider for L1DataProvider {
         number: L1BatchNumber,
         last_l2_block: &L2BlockHeader,
     ) -> TreeDataProviderResult {
+        let sl_chain_id = self
+            .pool
+            .connection_tagged("tree_data_fetcher")
+            .await
+            .map_err(|err| TreeDataFetcherError::Internal(err.into()))?
+            .eth_sender_dal()
+            .get_batch_commit_chain_id(number)
+            .await
+            .map_err(|err| TreeDataFetcherError::Internal(err.into()))?;
+        let chain_data = match sl_chain_id {
+            Some(chain_id) => {
+                let Some(chain_data) = self.chain_data_by_id(chain_id) else {
+                    return Err(TreeDataFetcherError::Internal(anyhow::anyhow!(
+                        "failed to find client for chain id {chain_id}"
+                    )));
+                };
+                chain_data
+            }
+            None => &self.l1_chain_data,
+        };
+
         let l1_batch_seal_timestamp = last_l2_block.timestamp;
         let from_block = self.past_l1_batch.and_then(|info| {
```

Comment:
Would it make sense to retain past L1 batches for both L1 and gateway chains (e.g., as a part of their data)? This could allow to not perform binary search below as often if `sl_chain_id` changes often. Or maybe, it makes sense to have to `L1DataProvider` instances for 2 recognized `sl_chain_id`s, e.g. by creating a new struct implementing `TreeDataProvider` that would route to these instances depending on the chain ID.

Reply:
wrt retaining I don't think it's needed, changing settlement layer is rare event (most of the chains will do it only once) so performing binary search is not a problem. Regarding new struct for routing I don't think it's really needed, IMO it's fine to keep routing logic as a part of `L1DataProvider` for now. But we definitely should consider adding it if we ever need to support >2 SLs